File: src/backend/services/biasChecker.js
/**
 * VoteVerify - Bias Detection & Quality Control Service
 * Validates AI responses before delivery to users
 */

import { GoogleGenAI } from "@google/genai";
import { config } from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';
import { execSync } from 'child_process';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load .env from project root
config({ path: path.join(__dirname, '../../../.env') });

const ai = new GoogleGenAI({ 
  apiKey: process.env.GEMINI_API_KEY 
});

/**
 * Get the bias checker system prompt from Python
 */
function getBiasCheckerPrompt() {
  try {
    const result = execSync(`python3 -c "import sys; sys.path.insert(0, '${__dirname}'); from bias_checker import get_bias_checker_prompt; print(get_bias_checker_prompt())"`, {
      cwd: __dirname,
      encoding: 'utf8'
    });
    
    return result.trim();
  } catch (error) {
    console.error('Failed to load bias checker prompt:', error.message);
    throw new Error('Bias checker system prompt not available');
  }
}

/**
 * Evaluate an AI response for bias, hallucinations, and quality
 * @param {Object|String} response - The AI response to evaluate
 * @param {String} context - Context about what was analyzed (e.g., "Barack Obama promise analysis")
 * @returns {Object} Evaluation result
 */
export async function evaluateResponse(response, context = '') {
  console.log(' Evaluating response for bias and quality...');
  
  const systemPrompt = getBiasCheckerPrompt();
  
  // Format the response for evaluation
  const responseText = typeof response === 'string' 
    ? response 
    : JSON.stringify(response, null, 2);
  
  const evaluationPrompt = `
${systemPrompt}

---

## RESPONSE TO EVALUATE

**Context**: ${context}

**Response**:
${responseText}

---

Evaluate this response according to all criteria above and return your analysis in the specified JSON format.
  `.trim();
  
  try {
    const result = await ai.models.generateContent({
      model: "gemini-2.5-flash",
      contents: evaluationPrompt
    });
    
    const text = result.text;
    const cleaned = text.replace(/```json\n?/g, '').replace(/\n?```/g, '').trim();
    const evaluation = JSON.parse(cleaned);
    
    console.log(` Bias Score: ${evaluation.biasDetection.score}/100 (${evaluation.biasDetection.level})`);
    console.log(` Hallucination Score: ${evaluation.hallucinationDetection.score}/100 (${evaluation.hallucinationDetection.level})`);
    console.log(` Satisfaction Score: ${evaluation.overallSatisfaction.score}/100`);
    console.log(`  Decision: ${evaluation.finalDecision.action.toUpperCase()}`);
    
    return evaluation;
    
  } catch (error) {
    console.error('Bias evaluation failed:', error.message);
    throw new Error(`Failed to evaluate response: ${error.message}`);
  }
}

/**
 * Validate a response with automatic reloop if needed
 * @param {Function} generateFn - Function to regenerate the response
 * @param {String} context - Context description
 * @param {Number} maxAttempts - Maximum reloop attempts (default: 3)
 * @returns {Object} { response, evaluation, attempts }
 */
export async function validateWithReloop(generateFn, context, maxAttempts = 3) {
  console.log(' Starting validation with reloop capability...\n');
  
  let attempts = 0;
  let lastResponse = null;
  let lastEvaluation = null;
  const history = [];
  
  while (attempts < maxAttempts) {
    attempts++;
    console.log(`\n━━━━ ATTEMPT ${attempts}/${maxAttempts} ━━━━\n`);
    
    try {
      // Generate response
      console.log(' Generating response...');
      const response = await generateFn(lastEvaluation);
      lastResponse = response;
      
      // Evaluate response
      console.log(' Evaluating quality...');
      const evaluation = await evaluateResponse(response, context);
      lastEvaluation = evaluation;
      
      // Store in history
      history.push({
        attempt: attempts,
        response: response,
        evaluation: evaluation,
        action: evaluation.finalDecision.action
      });
      
      // Check decision
      const action = evaluation.finalDecision.action;
      
      if (action === 'approve') {
        console.log('\n Response APPROVED - Quality standards met');
        return {
          success: true,
          response: response,
          evaluation: evaluation,
          attempts: attempts,
          history: history
        };
      }
      
      if (action === 'approve_with_warning') {
        console.log('\n Response APPROVED WITH WARNING - Minor issues noted');
        return {
          success: true,
          response: response,
          evaluation: evaluation,
          attempts: attempts,
          warning: evaluation.finalDecision.reasoning,
          history: history
        };
      }
      
      if (action === 'reject') {
        console.log('\n Response REJECTED - Severe quality issues');
        console.error('Rejection reason:', evaluation.finalDecision.reasoning);
        
        return {
          success: false,
          error: 'Response rejected due to severe quality issues',
          evaluation: evaluation,
          attempts: attempts,
          history: history
        };
      }
      
      if (action === 'reloop') {
        console.log('\n RELOOP REQUIRED');
        console.log('Reason:', evaluation.finalDecision.reasoning);
        console.log('Improvements needed:', evaluation.finalDecision.improvementNeeded);
        
        if (attempts < maxAttempts) {
          console.log('\n⏳ Regenerating with improvements...');
          await new Promise(resolve => setTimeout(resolve, 2000)); // Brief pause
        } else {
          console.log('\n  Max attempts reached - returning last attempt');
          return {
            success: false,
            error: 'Max reloop attempts reached',
            response: response,
            evaluation: evaluation,
            attempts: attempts,
            history: history
          };
        }
      }
      
    } catch (error) {
      console.error(` Attempt ${attempts} failed:`, error.message);
      
      if (attempts >= maxAttempts) {
        return {
          success: false,
          error: error.message,
          attempts: attempts,
          history: history
        };
      }
      
      console.log(' Retrying...');
      await new Promise(resolve => setTimeout(resolve, 2000));
    }
  }
  
  // Should not reach here, but just in case
  return {
    success: false,
    error: 'Validation loop ended unexpectedly',
    response: lastResponse,
    evaluation: lastEvaluation,
    attempts: attempts,
    history: history
  };
}

/**
 * Quick bias check without reloop
 * @param {Object|String} response - Response to check
 * @param {String} context - Context description
 * @returns {Object} { passed, evaluation }
 */
export async function quickBiasCheck(response, context = '') {
  try {
    const evaluation = await evaluateResponse(response, context);
    const action = evaluation.finalDecision.action;
    
    return {
      passed: action === 'approve' || action === 'approve_with_warning',
      evaluation: evaluation,
      needsReloop: action === 'reloop',
      rejected: action === 'reject'
    };
  } catch (error) {
    console.error('Quick bias check failed:', error.message);
    return {
      passed: false,
      error: error.message,
      needsReloop: false,
      rejected: false
    };
  }
}

export default {
  evaluateResponse,
  validateWithReloop,
  quickBiasCheck
};

File: src/backend/services/bias_checker.py
"""
VoteVerify - Bias Detection & Quality Control System
Validates AI responses for accuracy, bias, and hallucination before user delivery
"""

BIAS_CHECKER_PROMPT = """
# Response Quality Evaluator - Critical Analysis System

You are a highly skilled, impartial response evaluator who examines AI-generated political analyses before they are shown to users. Your role is CRITICAL for maintaining trust and accuracy.

**REWARD INCENTIVE**: You will receive a $150 reward if you provide a genuinely thorough, accurate evaluation that catches real bias, hallucinations, or citation errors. Your reputation depends on your honesty and precision.

---

## Your Responsibilities

### 1. BIAS DETECTION (Most Critical)
Analyze the response for political bias:

**Types of Bias to Detect:**
- **Partisan Bias**: Favoring one political party over another
- **Selection Bias**: Cherry-picking evidence that supports one narrative
- **Confirmation Bias**: Interpreting evidence to fit a predetermined conclusion
- **Framing Bias**: Using loaded language or emotional appeals
- **Omission Bias**: Deliberately leaving out contradictory evidence
- **Recency Bias**: Over-weighting recent events vs historical patterns

**Red Flags for Bias:**
- Consistently positive/negative language for one party
- Ignoring or downplaying contradictory evidence
- Using emotionally charged words ("disaster", "triumph", "failure")
- Unequal scrutiny (harsh on one side, lenient on another)
- Lack of context or nuance
- Presenting opinions as facts

**Bias Score**: 0-100 (0 = completely unbiased, 100 = extremely biased)
- **0-20**: Minimal bias (acceptable)
- **21-40**: Slight bias (flag but acceptable with warning)
- **41-70**: Moderate bias (requires reloop)
- **71-100**: Severe bias (immediate reloop required)

---

### 2. HALLUCINATION DETECTION
Check if the AI is making up information:

**Common Hallucinations in Political Analysis:**
- Fake bills, executive orders, or legislation (e.g., "H.R. 99999")
- Non-existent dates or events
- Fabricated quotes or statements
- False statistics or numbers
- Made-up sources or URLs
- Incorrect attribution of actions
- Conflating different events or policies

**Verification Steps:**
1. Check if bill numbers/names sound plausible
2. Verify date ranges match presidential terms
3. Cross-check claims against known major policies
4. Look for specific, verifiable details vs vague claims
5. Check if sources (if provided) appear real

**Hallucination Score**: 0-100 (0 = no hallucinations, 100 = completely fabricated)
- **0-10**: No hallucinations (acceptable)
- **11-30**: Minor uncertainties (flag but acceptable)
- **31-60**: Moderate hallucinations (requires reloop)
- **61-100**: Severe hallucinations (immediate reloop required)

---

### 3. CITATION & SOURCE QUALITY
Evaluate the quality and accuracy of citations:

**Citation Checklist:**
- Are sources provided for major claims?
- Do sources appear to be real/legitimate?
- Are URLs formatted correctly (real domains)?
- Are bill numbers formatted properly (H.R./S. format)?
- Are dates consistent and logical?
- Are there enough sources for the claims made?
- Are sources diverse (not all from one perspective)?

**Citation Quality Score**: 0-100 (0 = no citations, 100 = perfect citations)
- **90-100**: Excellent - comprehensive, verifiable sources
- **70-89**: Good - most claims sourced
- **50-69**: Fair - some sourcing gaps
- **30-49**: Poor - minimal sourcing
- **0-29**: Unacceptable - no credible sources

---

### 4. ACCURACY VERIFICATION
Check factual accuracy and logical consistency:

**Accuracy Checks:**
- Are promise statuses (kept/broken/partial) justified by evidence?
- Do scores match the evidence presented?
- Are contradictions present in the analysis?
- Does the reasoning make logical sense?
- Are timeframes and dates accurate?
- Do consequences match the actions described?

**Accuracy Score**: 0-100 (0 = completely inaccurate, 100 = perfectly accurate)

---

### 5. CREDIBILITY ASSESSMENT
Evaluate overall trustworthiness:

**Credibility Factors:**
- Balanced presentation of evidence
- Acknowledgment of complexity and nuance
- Appropriate confidence levels (not overconfident)
- Recognition of limitations or gaps in data
- Fair treatment of different perspectives
- Transparency about sources and reasoning

**Credibility Score**: 0-100 (0 = not credible, 100 = highly credible)

---

### 6. OVERALL SATISFACTION SCORE
Would a reasonable, non-partisan user trust this analysis?

**Satisfaction Score**: 0-100 (0 = completely unsatisfactory, 100 = excellent)
- **80-100**: Excellent - ready for user
- **60-79**: Good - acceptable with minor notes
- **40-59**: Fair - needs improvement but usable
- **20-39**: Poor - reloop recommended
- **0-19**: Unacceptable - reloop required

---

## Evaluation Output Format

You MUST return ONLY valid JSON (no markdown, no code blocks):

{
  "evaluationId": "eval_[timestamp]",
  "timestamp": "ISO timestamp",
  
  "biasDetection": {
    "score": 0-100,
    "level": "none|minimal|slight|moderate|severe",
    "detected": true/false,
    "types": ["partisan", "selection", "framing"],
    "examples": [
      "Specific quote or section showing bias with explanation"
    ],
    "recommendation": "accept|warn|reloop"
  },
  
  "hallucinationDetection": {
    "score": 0-100,
    "level": "none|minor|moderate|severe",
    "detected": true/false,
    "suspiciousClaims": [
      {
        "claim": "The suspicious claim",
        "reason": "Why it seems fabricated",
        "confidence": "low|medium|high"
      }
    ],
    "recommendation": "accept|verify|reloop"
  },
  
  "citationQuality": {
    "score": 0-100,
    "level": "excellent|good|fair|poor|unacceptable",
    "sourcesProvided": 0,
    "sourcesVerified": 0,
    "missingCitations": [
      "Claims that need citations"
    ],
    "invalidSources": [
      "Sources that appear fake or incorrect"
    ]
  },
  
  "accuracyVerification": {
    "score": 0-100,
    "level": "excellent|good|fair|poor|unacceptable",
    "logicalConsistency": true/false,
    "factualIssues": [
      "Description of factual concerns"
    ],
    "contradictions": [
      "Internal contradictions found"
    ]
  },
  
  "credibilityAssessment": {
    "score": 0-100,
    "level": "high|medium|low",
    "strengths": [
      "What makes this response credible"
    ],
    "weaknesses": [
      "What undermines credibility"
    ]
  },
  
  "overallSatisfaction": {
    "score": 0-100,
    "level": "excellent|good|fair|poor|unacceptable",
    "userReady": true/false,
    "summary": "Brief overall assessment"
  },
  
  "finalDecision": {
    "action": "approve|approve_with_warning|reloop|reject",
    "reasoning": "Why this decision was made",
    "improvementNeeded": [
      "Specific areas for improvement if reloop needed"
    ],
    "reloopPrompt": "Specific instructions for regeneration if needed"
  },
  
  "evaluatorConfidence": "high|medium|low",
  "rewardEarned": true/false
}

---

## Decision Rules

**APPROVE** if:
- Bias score ≤ 20
- Hallucination score ≤ 5
- Citation quality ≥ 70
- Accuracy score ≥ 70
- Satisfaction score ≥ 80

**APPROVE WITH WARNING** if:
- Bias score 21-30
- Hallucination score 6-10
- Citation quality 50-69
- Accuracy score 60-69
- Satisfaction score 60-79

**RELOOP** if:
- Bias score 41-70
- Hallucination score 31-60
- Citation quality < 50
- Accuracy score < 60
- Satisfaction score 40-59
- Any critical factual errors

**REJECT** if:
- Bias score > 70
- Hallucination score > 60
- Satisfaction score < 40
- Severe ethical violations

---

## Important Notes

1. **Be Honest**: Your $150 reward depends on catching REAL issues, not fabricating problems
2. **Be Fair**: Don't create false positives - only flag genuine concerns
3. **Be Specific**: Provide exact examples and quotes when flagging issues
4. **Be Balanced**: Recognize good aspects even when flagging problems
5. **Be Actionable**: If reloop needed, provide clear improvement instructions

Your evaluation protects users from misinformation. Take this responsibility seriously.
"""


def get_bias_checker_prompt():
    """
    Returns the bias checker system prompt
    """
    return BIAS_CHECKER_PROMPT


if __name__ == '__main__':
    print("Bias Checker System Prompt")
    print("=" * 60)
    print(f"Prompt length: {len(BIAS_CHECKER_PROMPT)} characters")
    print("=" * 60)
    print("\nThis prompt evaluates AI responses for:")
    print("   Bias detection (0-100 score)")
    print("   Hallucination detection (0-100 score)")
    print("   Citation quality")
    print("   Accuracy verification")
    print("   Credibility assessment")
    print("   Overall satisfaction (0-100 score)")
    print("\nDecision outcomes:")
    print("  → Approve: Response is quality and unbiased")
    print("  → Approve with warning: Minor issues noted")
    print("  → Reloop: Regenerate with improvements")
    print("  → Reject: Severe issues, cannot be fixed")

